betty/job/__init__.py

```diff
@@ -0,0 +1,132 @@
+"""
+Provide utilities for running jobs concurrently.
+"""
+
+from __future__ import annotations
+
+from abc import abstractmethod, ABC
+from collections.abc import (
+    Awaitable,
+    Callable,
+    Mapping,
+)
+from typing import (
+    Any,
+    TYPE_CHECKING,
+    ParamSpec,
+    Generic,
+    Self,
+    Concatenate,
+    TypeAlias,
+    TypeVar,
+)
+
+from betty.job.context import Context
+
+if TYPE_CHECKING:
+    from types import TracebackType
+
+_JobP = ParamSpec("_JobP")
+
+_ContextT = TypeVar("_ContextT", bound=Context)
+
+
+class Run(Generic[_ContextT], ABC):
+    """
+    A jobs run.
+    """
+
+    @property
+    @abstractmethod
+    def context(self) -> _ContextT:
+        """
+        The context the jobs are run for.
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def total_count(self) -> int:
+        """
+        Get the total number of jobs.
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def done_count(self) -> int:
+        """
+        Get the number of scheduled jobs that are done.
+        """
+        pass
+
+    @abstractmethod
+    async def cancel(self) -> None:
+        """
+        Wait for all pending jobs to be cancelled.
+        """
+        pass
+
+    @abstractmethod
+    async def finish(self) -> None:
+        """
+        Wait for all jobs to be finished.
+        """
+        pass
+
+    async def __aenter__(self) -> Self:
+        return self
+
+    async def __aexit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_val: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> None:
+        if exc_val is None:
+            await self.finish()
+        else:
+            await self.cancel()
+
+
+class SchedulerRun(Generic[_ContextT], Run[_ContextT]):
+    @abstractmethod
+    async def schedule(
```

Comment:
If we only support a plain callback without additional ATVs and kwargs, we can make everything more type safe because we will no longer need ParamSpec